src/vm/object.cpp

```diff
@@ -2051,49 +2051,45 @@ StackTraceElement & StackTraceArray::operator[](size_t index)
 }
 
 #if !defined(DACCESS_COMPILE)
-// Define the lock used to access stacktrace from an exception object
-SpinLock g_StackTraceArrayLock;
 
 void ExceptionObject::SetStackTrace(StackTraceArray const & stackTrace, PTRARRAYREF dynamicMethodArray)
 {        
     CONTRACTL
     {
-        GC_NOTRIGGER;
-        NOTHROW;
+        GC_TRIGGERS;
+        THROWS;
         MODE_COOPERATIVE;
     }
     CONTRACTL_END;
 
-    Thread *m_pThread = GetThread();
-    SpinLock::AcquireLock(&g_StackTraceArrayLock, SPINLOCK_THREAD_PARAM_ONLY_IN_SOME_BUILDS);
+    _dispatchStateLock->EnterObjMonitor();
 
     SetObjectReference((OBJECTREF*)&_stackTrace, (OBJECTREF)stackTrace.Get());
     SetObjectReference((OBJECTREF*)&_dynamicMethods, (OBJECTREF)dynamicMethodArray);
 
-    SpinLock::ReleaseLock(&g_StackTraceArrayLock, SPINLOCK_THREAD_PARAM_ONLY_IN_SOME_BUILDS);
+    _dispatchStateLock->LeaveObjMonitor();
 
 }
 
 void ExceptionObject::SetNullStackTrace()
 {        
     CONTRACTL
     {
-        GC_NOTRIGGER;
-        NOTHROW;
+        GC_TRIGGERS;
```

Comment:
Flipping `GC_NOTRIGGER` to `GC_TRIGGERS` requires careful review of both the implementation and all callers to make sure that all GC references are reported to the GC. It is likely what's causing the crashes you are seeing.

Reply:
Will go through it again but the `stackFramesData` pointer passed from `ReportExceptionStackHelper` looked to be the only one (which becomes a `PTRARRAYREF* outDynamicMethodArray` param with `*outDynamicMethodArray = _dynamicMethods` assignment) so I've added that; however it itself is defined in a `GC_TRIGGERS` method that uses `GCPROTECT_BEGIN` already; so I might be missing something on how this works?

Reply:
https://github.com/dotnet/coreclr/blob/master/Documentation/coding-guidelines/clr-code-guide.md#212-your-first-gc-hole is the doc that tries to explain how this works.

Reply:
Ah! I understand COOPERATIVE mode now; and it would have worked before because it was a NOGC method